web/server/network_check.go

```diff
@@ -0,0 +1,326 @@
+func testGeneral(ctx context.Context) bool {
+	timeout := 5 * time.Second
+	attempts := 1
+	if runtime.GOOS == "windows" {
+		// TODO(RSDK-8344): this is temporary as we 1) debug connectivity issues on windows,
+		// and 2) migrate to using the native checks on the underlying connection.
+		timeout = 15 * time.Second
+		attempts = 2
+	}
+
+	for i := range attempts {
+		// Use DialDirectGRPC to make a connection to app.viam.com instead of a
+		// basic net.Dial in order to ensure that the connection can be made
+		// behind wifi or the BLE-SOCKS bridge (DialDirectGRPC can dial through
+		// the BLE-SOCKS bridge.)
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		conn, err := rpc.DialDirectGRPC(ctx, "app.viam.com:443", nil)
+		cancel()
+		if err == nil {
+			conn.Close() //nolint:gosec,errcheck
+			return true
+		}
+		if i < attempts-1 {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return false
+}
+
+var (
+	stunServerURLsToTestUDP = []string{
+		"global.stun.twilio.com:3478",
+		"turn.viam.com:443",
+		"stun.l.google.com:3478",
+		"stun.l.google.com:19302",
+		"stun.sipgate.net:3478",
+		"stun.sipgate.net:3479",
+	}
+	stunServerURLsToTestTCP = []string{
+		"turn.viam.com:443", // only STUN server that acceps TCP STUN traffic
+	}
+)
+
+// Tests NAT over UDP against STUN servers.
+func testUDP(ctx context.Context, logger logging.Logger) error {
+	// Listen on arbitrary UDP port.
+	conn, err := net.ListenPacket("udp", "0.0.0.0:0")
+	if err != nil {
+		logger.Warn("Failed to listen over UDP on a port; UDP traffic may be blocked")
+		return err
+	}
+
+	// `net.PacketConn`s do not function with contexts (only deadlines.) If passed-in
+	// context expires (machine is likely shutting down,) _or_ tests finish, close the
+	// underlying `net.PacketConn` asynchronously to stop ongoing network checks.
+	testUDPDone := make(chan struct{})
+	defer func() {
+		testUDPDone <- struct{}{}
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-testUDPDone:
+		}
+		conn.Close() //nolint:gosec,errcheck
+	}()
+
+	// Set a deadline for all interactions of 10 seconds in the future.
+	if err := conn.SetDeadline(time.Now().Add(10 * time.Second)); err != nil {
+		return err
+	}
+
+	// Build a STUN binding request to be used against all STUN servers.
+	bindRequest, err := stun.Build([]stun.Setter{
+		stun.TransactionID,
+		stun.BindingRequest,
+	}...)
+	if err != nil {
+		return err
+	}
+	bindRequestRaw, err := bindRequest.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	var stunResponses []*STUNResponse
+	for _, stunServerURLToTest := range stunServerURLsToTestUDP {
+		if ctx.Err() != nil {
+			logger.Info("Machine shutdown detected; stopping UDP network tests")
+			return nil
+		}
+
+		logger := logger.WithFields("stun_server_url", stunServerURLToTest)
+
+		stunResponse := NewSTUNResponse(stunServerURLToTest)
+		stunResponses = append(stunResponses, stunResponse)
+
+		udpAddr, err := net.ResolveUDPAddr("udp4", stunServerURLToTest)
```

Comment:
I think we need to be smarter about isolating dns errors with message request/response errors. Such that even if we get a dns error, we can try some of the raw IPs we expect to be running.

There'd be a risk that our raw IP mapping goes stale -- we should prefer* what the dns resolver gives us when it does work.

Given this is a bit nebulous, I think punting and keeping what you have here is best, in the interest of getting this out sooner. I just know that the DNS thing was annoying to workaround back when we had our stun server running, but it didn't have a dns entry yet.